src/entrypoint/graphql-admin-server.ts

```diff
@@ -0,0 +1,187 @@
+import { ApolloServer } from "apollo-server-express"
+import dotenv from "dotenv"
+import express from "express"
+import expressJwt from "express-jwt"
+import { applyMiddleware } from "graphql-middleware"
+import { and, rule, shield } from "graphql-shield"
+import mongoose from "mongoose"
+import pino from "pino"
+import PinoHttp from "pino-http"
+import { v4 as uuidv4 } from "uuid"
+
+import { yamlConfig } from "../config"
+import { AuthorizationError, IPBlacklistedError } from "../error"
+import { baseLogger } from "../logger"
+import { redis } from "../redis"
+import { User } from "../schema"
+import { setupMongoConnection } from "../mongodb"
+import helmet from "helmet"
+import { isDev, updateIPDetails, isIPBlacklisted } from "../utils"
+import { WalletFactory } from "../walletFactory"
+import { gqlAdminSchema } from "../graphql"
+
+dotenv.config()
+
+const graphqlLogger = baseLogger.child({ module: "graphql" })
+
+// TODO: extract common server configs
+
+const pino_http = PinoHttp({
+  logger: graphqlLogger,
+  wrapSerializers: false,
+
+  // Define custom serializers
+  serializers: {
+    err: pino.stdSerializers.err,
+    req: pino.stdSerializers.req,
+    res: (res) => ({
+      // FIXME: kind of a hack. body should be in in req. but have not being able to do it.
+      body: res.req.body,
+      ...pino.stdSerializers.res(res),
+    }),
+  },
+  autoLogging: {
+    ignorePaths: ["/healthz"],
+  },
+})
+
+const isAuthenticated = rule({ cache: "contextual" })(async (parent, args, ctx) => {
+  if (ctx.uid === null) {
+    throw new AuthorizationError(undefined, {
+      logger: graphqlLogger,
+      request: ctx.request.body,
+    })
+  }
+  return true // ?
+})
+
+const isEditor = rule({ cache: "contextual" })(async (parent, args, ctx) => {
+  return ctx.user.role === "editor"
+})
+
+const permissions = shield(
+  {
+    Query: {
+      allLevels: and(isAuthenticated, isEditor),
+      userDetailsByPhone: and(isAuthenticated, isEditor),
+      userDetailsByUsername: and(isAuthenticated, isEditor),
+    },
+    Mutation: {
+      // currentTime: and(isAuthenticated, isEditor),
+    },
+  },
+  { allowExternalErrors: true },
+)
+
+export async function startApolloServerForAdminSchema() {
+  const app = express()
+
+  const schemaWithPermissions = applyMiddleware(gqlAdminSchema, permissions)
```

Comment:
beyond the `gqlAdminSchema` parameter, anything else seems to be a copy of the main graphql. 

I'm wondering if you should not just create a function instead that is being used by both the main graphql and the admin graphql to avoid the duplication?